Add console.group support

nchevobbe opened the pull request:
Fixes #44

Test instructions : 
1. `npm test`

@linclark @bgrins I think this is working pretty well like this. I wanted to have your opinion about it before the weekend, so I can change what needs to be, and write the tests. 

The main ideas here are: 
- we add a `group` property on the state, which is a map which contains the parent groups of a group
- we add a `currentGroup` in order to assign the id of the latest created group in a `groupId` on the message
- we re-use the existing `messagesUiById` to track the open/closed state of the group
- in the selector, we only return the message which are in an opened group
- when pruning, if there is one, we output the last non-closed group

Since I added a lot of thing in the messages selector, I revamped how we do the filtering to something simpler to understand IMO.

I tweaked the design a little bit, but I think we could have a follow-up with helen's take on this. 

And that's it :) Please tell me what you think about this, I'll do my best to have it done over the week-end. Thanks !

linclark commented:
\o/

I'm going to get another PR up, but after that I should be able to take a look

linclark commented on devtools/client/themes/webconsole.css:
```diff
@@ -708,6 +708,11 @@ a.learn-more-link.webconsole-learn-more-link {
   flex: 1 1 100%;
 }
 
+.message[class*="startGroup"] .message-body {
```
Is startGroup used anywhere? I only see it in the CSS.

nchevobbe replied:
Since the PR introducing the generic `Message` component, it's automatically added to the classes, because it's the level ( there is `startGroup` and `startGroupCollapsed` ). I don't know if we'd better add the existing `group` className instead of those two.

linclark replied:
ah, I see. We might want to avoid using the wildcard here. I know that complex selectors can have an impact on performance. @bgrins might know more about that

linclark commented on devtools/client/webconsole/new-console-output/constants.js:
```diff
@@ -16,6 +16,8 @@ const actionTypes = {
   FILTER_TEXT_SET: "FILTER_TEXT_SET",
   FILTERS_CLEAR: "FILTERS_CLEAR",
   FILTER_BAR_TOGGLE: "FILTER_BAR_TOGGLE",
+  GROUP_OPEN: Symbol(),
```
Hm, why Symbol?

nchevobbe replied:
Oh, I read about it and wanted to try it out, since is looks like a good candidate for such thing ( we never actually check the `actionTypes` string, but always against the constant, so having a unique value you can't clash seems like a good idea). Anyway, it was for test purpose and I'll remove it. If using symbols looks like a good idea, I might make the change in a future PR (but it might not be a good idea ? maybe we do want to have those strings ? )

linclark replied:
Ah, interesting. I'd say that since most documentation uses strings, it's less confusing to people new to the code base if we also use strings. But it is an interesting idea.

nchevobbe replied:
BTW, those action are useless since we are using MESSAGE_OPEN and MESSAGE_CLOSE for this

linclark commented:
One thing I notice in testing this out: evaluation results aren't indented. 

<img width="228" alt="nightly_start_page" src="https://cloud.githubusercontent.com/assets/105464/19014475/90dfb8b0-87bb-11e6-916f-9a225a62fd39.png">

I'm not sure whether that's intentional, or if it's just because this is still WIP. We probably should indent them like Chrome does

![image](https://cloud.githubusercontent.com/assets/105464/19014482/cdcf48b2-87bb-11e6-803b-6af8ed6ff0a1.png)

nchevobbe commented:
> I'm not sure whether that's intentional

That's not, I must be doing something wrong. Thanks for seeing that

linclark commented on devtools/client/webconsole/new-console-output/reducers/messages.js:
```diff
@@ -12,19 +12,34 @@ const MessageState = Immutable.Record({
 
       if (newMessage.type === constants.MESSAGE_TYPE.NULL_MESSAGE) {
         return state;
+      } else if (newMessage.type === constants.MESSAGE_TYPE.END_GROUP) {
```
Could you add some comments to the code in this file? I think I understand what's going on here and it makes sense to me, but this is kind of complex and worth spelling out I think.

linclark commented on devtools/client/webconsole/new-console-output/selectors/messages.js:
```diff
@@ -14,22 +14,35 @@ const {
+          )
+        )
+      );
+    }),
```
I like the changes here :)

linclark commented on devtools/client/webconsole/new-console-output/selectors/messages.js:
```diff
@@ -113,7 +140,18 @@ function isTextInFrame(text, frame) {
-    return messages.splice(0, messageCount - logLimit);
+    // If the second non-pruned message is in a group,
+    // we want to return the group as the first non-pruned message.
+    let firstIndex = messages.size - logLimit;
```
I'm glad to see this. This was something that we knew was going to be tricky. There could still be edge cases in here, but I don't think `console.group` is so frequently used that we need to suss out those edge cases.

nchevobbe commented on devtools/client/webconsole/new-console-output/components/indent.js:
```diff
@@ -0,0 +1,35 @@
+  render: function () {
+    const { indent } = this.props;
+
+    // if (indent > 0) {
```
I don't know yet. At first it seemed right for me to not display anything superfluous, but now i think it is easier to have the same skeleton for all messages

nchevobbe commented on devtools/client/webconsole/new-console-output/components/message.js:
```diff
@@ -78,9 +82,14 @@ function Message(props) {
   // If there is an expandable part, make it collapsible.
   let collapse = null;
-  if (attachment) {
+  if (attachment || isGroupType(type)) {
```
we didn't have one in the current console, but we have the tooltip for the stacktrace for example, and I think we should be consistent.

nchevobbe commented:
@linclark @bgrins I just pushed 2 commits (https://github.com/devtools-html/gecko-dev/pull/339#commits-pushed-5adcb18) that I think are ready for review.
I did some modifications based on Lin's comments and added tests.

Even if the look of the console.group is a bit raw, I think we can handle it as a follow up later, when Helen have some times to think about it.

Thanks !

linclark commented on devtools/client/webconsole/new-console-output/test/store/messages.test.js:
```diff
@@ -131,6 +133,75 @@ describe("Message reducer:", () => {
+      expect(messages.size).toBe(0);
+    });
+
+    it("does not add message after a console.groupCollapsed message", () => {
```
I just want to clarify... it does add the message, but doesn't display it, right?

nchevobbe commented:
> One thing I noticed... there doesn't seem to be any testing that indent levels are correct. I wouldn't block on it, but that would be a nice to have. 

yes, I was thinking about this this morning, I'll add tests for all of the message types.

> We also might want to introduce a single mochitest just to test the happy path... but again, that's not a blocker.

I don't mind adding that to this PR.

nchevobbe commented:
@linclark I think everything is okay for review here

linclark commented on devtools/client/webconsole/new-console-output/test/mochitest/browser_webconsole_console_group.js:
```diff
@@ -0,0 +1,127 @@
+  });
+});
+
+function findConsoleMessage(node, indent = 0, index) {
```
I might call this `findIndentedMessage`. That seems to be the major difference between it and the `findMessage` declared in head.

nchevobbe commented on devtools/client/webconsole/new-console-output/components/message-indent.js:
```diff
@@ -23,7 +23,12 @@ const MessageIndent = createClass({
+      className: "indent",
+      style: {"--indent": indent},
+      // For testing purpose
+      "data-indent": indent,
```
I'm not sure I understand.
Are you talking about a CSS selector ( as opposed to a Redux one) ?
We don't have nesting here, there is always only one `span.indent` , which we are sizing according to the css variable `--indent` on the element and AFAIK there's no way I can include that in a selector. 
We could test the css variable value for the element, but I feel like this is worse than checking an attribute.

Again, I may have misunderstood your point, so feel free to tell me if I'm wrong :)

linclark replied:
Ah, sorry, sick brain wasn't looking closely enough. Why do you think testing the css variable value would be worse than checking an attribute?

nchevobbe replied:
IMO it's easier since we can target it directly in a selector `.message .indent[data-indent=2]` , instead of getting the node, and then doing `node.style.getPropertyValue("--indent")` , but it may be just me.

Secondly, React adds a `px` suffix to the css variable (https://facebook.github.io/react/tips/style-props-value-px.html) There's nothing we can do to prevent that, so it means that we would have to test `node.style.getPropertyValue("--indent").toBe("2px")` , which is really not great.

That said, maybe having a css variable for this might not be the good choice ? What I was really aiming for in the first place was to directly use `data-indent` to dynamically compute the width in the CSS, but that's not possible for now. 
We could compute directly the width in the `MessageIndent` component but it would give us less flexibility to tweak the style of the groups (we would have to modify the component itself, and we may want to compute other properties based on the variable), and this is kind of a blocker for me.

So, what do you think about all this ? keeping the data-indent or checking the variable or using something else ?

linclark replied:
> We could compute directly the width in the MessageIndent component

To be honest, I think this might be the best thing to do for now. I think less flexibility is good until we're ready to put the flexibility to use.

> There's nothing we can do to prevent that, so it means that we would have to test node.style.getPropertyValue("--indent").toBe("2px") , which is really not great.

What do you think is suboptimal about this? The pro that I see to doing it this way is that it is explicitly testing the thing that we expect to be in the output. The con that I see is that it will then make the test fail if React changes the way it does things.

nchevobbe commented:
@linclark I rebased and did the modifications we talked about, so everything shoule be good now.

I added more commit, instead of amending the old ones since I assume you're gonna squash them. But if you want to me to better organize them, I can do it.

linclark commented:
It looks like this adds the proptype for tableData back in, which makes tests fail because of this:

```
309 INFO TEST-UNEXPECTED-FAIL | devtools/client/webconsole/new-console-output/test/mochitest/browser_webconsole_console_table.js | Warning: Failed propType: Invalid prop `tableData` of type `object` supplied to `ConsoleApiCall`, expected `array`. Check the render method of `MessageContainer`. -
```

FYI, the iframe test is also failing right now, but that is independent of this bug (and related to an issue Brian already pointed out.

The npm tests are also failing consistently for me on this branch. Have you double checked the logic to make sure that it should pass? It could be an intermittent, but I'm wondering whether this isn't an actual error.

nchevobbe commented:
Ugh, sorry about that, I'm currently building so I can check all this.
For the filter test, I did check, but I'll check again to make sure what's going on.

linclark commented:
Awesome, thank you :)

nchevobbe commented:
> The npm tests are also failing consistently for me on this branch. Have you double checked the logic to make sure that it should pass? It could be an intermittent, but I'm wondering whether this isn't an actual error.

@linclark  I'm so dumb  :flushed: There was a typo in the network filtering function. I was trying to get `filters.network`, but the filter is `filter.net`. I should have check before opening a bug, sorry. This is fixed in https://github.com/devtools-html/gecko-dev/pull/339/commits/40d8b513a7959cd757bad376189e821e5df558c3 , the npm tests now pass.

I removed the tableData proptype in https://github.com/devtools-html/gecko-dev/pull/339/commits/c3ecd6f5b38d41ea3481e07198e15fe29411c0cc , and all the mochitest, except the frames one, now pass.

I checked the whole review, I hope I didn't miss anything.

linclark commented:
Ha, no worries. If that makes you dumb, then I'm dumb too... I make those mistakes all the time.

I manually tested and ran the tests... this is looking 💯

Thanks so much for all of your work on this. It's awesome to have a working `console.group()`. We're so very close to feature parity on console api!

nchevobbe commented:
I'm so happy this is merged :raised_hands:
